contracts/RewardByBlock.sol

```diff
@@ -79,26 +94,48 @@ contract RewardByBlock is EternalStorage, IRewardByBlock {
 
         uint256 i;
         
-        for (i = 0; i < extraLength; i++) {
-            uint256 extraIndex = i.add(2);
+        for (i = 0; i < length; i++) {
             address extraAddress = extraReceivers(i);
             uint256 extraAmount = extraReceiversAmounts(extraAddress);
             _setExtraReceiverAmount(0, extraAddress);
-            receivers[extraIndex] = extraAddress;
-            rewards[extraIndex] = extraAmount;
+            receivers[i.add(2)] = extraAddress;
+            rewards[i.add(2)] = extraAmount;
         }
 
         for (i = 0; i < receivers.length; i++) {
             _setMinted(rewards[i], receivers[i]);
         }
 
+        length = bridgesLength();
+        for (i = 0; i < length; i++) {
+            address bridgeAddress = bridges(i);
+            uint256 bridgeAmount = bridgesAmounts(bridgeAddress);
+            _setBridgeAmount(0, bridgeAddress);
+            _addMintedByBridge(bridgeAmount, bridgeAddress);
+        }
+
         _clearExtraReceivers();
+        _clearBridges();
 
         emit Rewarded(receivers, rewards);
     
         return (receivers, rewards);
     }
 
+    function bridges(uint256 _index) public view returns(address) {
```

Comment:
It means that besides adding the bridge address to `isBridgeContract` we need to call this method with the same address, doesn't it? I think such mechanism introduce a risk of data inconsistency. May we have `isBridgeContract()` that works with the same data which are used in `bridges()`.

For example,

```solidity
function isBridgeContract(address _addr) public pure returns(bool) {
    if (_addr == bridges(1)) return true;
    if (_addr == bridges(2)) return true;
    if (_addr == bridges(3)) return true;
    return false;
}

function bridges(uint256 _index) public view returns(address) {
    // These values must be changed before deploy
    if (_index == 1) return address(0x0000000000000000000000000000000000000000);
    if (_index == 2) return address(0x0000000000000000000000000000000000000000);
    if (_index == 3) return address(0x0000000000000000000000000000000000000000);
    return address(0);
}
```

So the `for`-loop in `reward()` could look as following:
```solidity
    i = 0;
    address bridgeAddress = bridges(0);
    while (bridgeAddress != address(0)) {
        uint256 bridgeAmount = bridgesAmounts(bridgeAddress);
        _setBridgeAmount(0, bridgeAddress);
        _addMintedByBridge(bridgeAmount, bridgeAddress);
        bridgeAddress = bridges(i++);
    }
```

Reply:
No, we don't need to call `bridges()` - this function is for internal use but I left it public just in case.

`bridges()` returns a bridge address by its index for the current block. This function is used by `reward` function.

When the bridge calls `addExtraReceiver`, its address is pushed in temporary array by `_addBridge` private function, because inside `reward` function we must know the list of the bridges that called `addExtraReceiver()`. Then we read bridges' addresses inside `reward` function to retrieve their amounts for statistics.

Every time when `reward` is called, it saves statistics for each participated bridge in the current block and then `_clearBridges` clears the current array of bridges to use the array from scratch in the next block.

Probably, the name of the `bridges` function is not very good or we could make it private.

Reply:
Another worry I have is the fact that every call of `_setBridgeAmount` requires two SSTORE operations. Could we optimize the code by having a map to store amounts for bridges? 
  * `_setBridgeAmount` will increase value of the particular element of the map (one SSTORE)
  * `_clearBridges` is not needed because the map values will be zeroed by `_setBridgeAmount(0, bridgeAddress);` in `reward()`
  * `reward` will read every element of the map to call `_addMintedByBridge`

It is just a suggestion for discussion.

Reply:
Yes, I'll try to use a hardcoded array of bridges' addresses instead. I'm working on that now and will show the final code after doing some tests.